### PR TITLE
Adds holt-winters query function

### PR DESCRIPTION
**Pull request by eliothedeman**
I messed up the rebase on my last PR... sorry.

### CODE REVIEW

**Comment by eliothedeman** [comment]
Heres an example.
7780b478-e581-11e5-9785-a012bd45aeeb

File: promql/bench.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import "testing"
+
+// A Benchmark holds context for running a unit test as a benchmark.
+type Benchmark struct {
+	b *testing.B
+	t *Test
+}
+
+// NewBenchmark returns an initilized empty Benchmark.
```
Comment (mdlayher) [inline promql/bench.go]: "initialized"

File: promql/bench.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		b.Fatalf("Unable to run benchmark: %s", err)
+	}
+	return &Benchmark{
+		b: b,
+		t: t,
+	}
+}
+
+// Run the benchmark.
```
Comment (mdlayher) [inline promql/bench.go]: "Run runs the benchmark."

**Comment by grobie** [comment]
Looks great! I haven't had time to test it yet though.

File: promql/functions.go [hunk 3]
```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
 	return resultVector
 }
 
+// Calculate the smoothed value at index i in raw data d.
```
Comment (brian-brazil) [inline promql/functions.go]: This comment is too low-level to understand if you don't already know what the function does

File: promql/functions.go [hunk 4]
```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+
+	// smoothing factor
+	sf := ev.evalFloat(args[1])
+
+	// trend factor
+	tf := ev.evalFloat(args[2])
+
+	// sanity check input
+	if sf <= 0 || sf >= 1 {
+		ev.errorf("invalid smoothing factor. Expected: 0 < tf < 1 got: %f", sf)
```
Comment (brian-brazil) [inline promql/functions.go]: s/tf/sf/

File: promql/functions.go [hunk 5]
```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+// has exponentially less influence on the current data. Holt-Winter also accounts
+// for trends in data. The smoothing factor (0 < sf < 1) effects how historical data will effect the current
+// data. A lower smoothing factor increases the influence of historical data. The trend factor (0 < tf < 1) effects
+// how trends in historical data will effect the current data. A higher trend factor increases the influence
+// of trends.
+// algorithm taken from https://en.wikipedia.org/wiki/Exponential_smoothing titled: "Double exponential smoothing"
+func funcHoltWinters(ev *evaluator, args Expressions) model.Value {
+	mat := ev.evalMatrix(args[0])
+
+	// smoothing factor
```
Comment (brian-brazil) [inline promql/functions.go]: Comments should begin with a capital letter and end with a full stop.

File: promql/functions.go [hunk 6]
```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+			d[i] = float64(v.Value)
+
+			// holt-winters is undefined for NaN
+			if math.IsNaN(d[i]) {
+				foundNaN = true
+				break
+			}
+		}
+
+		// skip to next because we found a NaN in the raw data set
```
Comment (brian-brazil) [inline promql/functions.go]: Usually we'd continue calculating, and let NaN fall out of the calculation

File: promql/functions.go [hunk 7]
```diff
@@ -184,6 +184,137 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
 }
 
+// Calculate the smoothed value at index i in raw data d.
+// The smoothed value is created by creating two values (x,y) where...
+// x = the current raw data point, weighted by the smoothing factor.
+// y = the last smoothed point + the last trend value weighted by the inverse of the smoothing factor.
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
```
Comment (brian-brazil) [inline promql/functions.go]: Having to track this seems odd for dynamic programming, you should already know what has been calculated.

For example the `doubleSVal(i-1)` below can be a straight lookup, and if you special case on `i` you should be able to remove the rest.

File: promql/functions.go (hunk 5 above)
Comment (eliothedeman) [inline promql/functions.go]: Should I not be matching the style of the comments of the other query functions?

File: promql/functions.go (hunk 6 above)
Comment (eliothedeman) [inline promql/functions.go]: NaN will turn all data points after it into NaN. I suppose we can remove it from the dataset and continue calculating?

Comment (brian-brazil) [inline promql/functions.go]: The way we're currently handling it would be to result in that having a NaN output

File: promql/functions.go (hunk 5 above)
Comment (juliusv) [inline promql/functions.go]: If you are referring to the odd `=== foo() ===` comments above PromQL function implementations, those are kind of a legacy exception. We normally start all other comments with capital letters and a period at the end. This helps legibility especially on multi-line / multi-sentence comments to clarify where a sentence begins and ends. But we do it everywhere else for consistency, too.

File: promql/functions.go [hunk 8]
```diff
@@ -184,6 +184,109 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+// Holt-Winters is similar to a weighted moving average, where historical data
+// has exponentially less influence on the current data. Holt-Winter also accounts
+// for trends in data. The smoothing factor (0 < sf < 1) effects how historical data will effect the current
+// data. A lower smoothing factor increases the influence of historical data. The trend factor (0 < tf < 1) effects
+// how trends in historical data will effect the current data. A higher trend factor increases the influence.
+// of trends. Algorithm taken from https://en.wikipedia.org/wiki/Exponential_smoothing titled: "Double exponential smoothing".
+func funcHoltWinters(ev *evaluator, args Expressions) model.Value {
+	mat := ev.evalMatrix(args[0])
+
+	// smoothing factor
```
Comment (brian-brazil) [inline promql/functions.go]: Comments should be sentences, start with a capital letter and end with a period.

File: promql/functions.go [hunk 9]
```diff
@@ -184,6 +184,109 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+
+	// sanity check input
+	if sf <= 0 || sf >= 1 {
+		ev.errorf("invalid smoothing factor. Expected: 0 < sf < 1 got: %f", sf)
+	}
+	if tf <= 0 || tf >= 1 {
+		ev.errorf("invalid trend factor. Expected: 0 < tf < 1 got: %f", sf)
+	}
+
+	// make an output vector large enough to hole the entire result
```
Comment (brian-brazil) [inline promql/functions.go]: hold

File: promql/functions.go [hunk 10]
```diff
@@ -184,6 +184,109 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+// x = the current raw data point, weighted by the smoothing factor.
+// y = the last smoothed point + the last trend value weighted by the inverse of the smoothing factor.
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	x := sf * d[i]
+	y := (1 - sf) * (s[i-1] + doubleBVal(i-1, sf, tf, s, b, d))
+
+	// cache value
+	s[i] = x + y
```
Comment (brian-brazil) [inline promql/functions.go]: This is duplicating what `funcHoltWinters` is doing.  I'd suggest moving that for loop into here, and just returning the result.

File: promql/functions.go (hunk 8 above)
Comment (eliothedeman) [inline promql/functions.go]: @brian-brazil I believe this comment already satisfies this. Am I missing something?

File: promql/test.go [hunk 11]
```diff
@@ -411,6 +412,30 @@ func (cmd clearCmd) String() string {
 	return "clear"
 }
 
+// RunAsBenchmark runs the test in benchmark mode.
+// This will not count any loads or non eval functions.
+func (t *Test) RunAsBenchmark(b *testing.B) error {
+	for _, cmd := range t.cmds {
+
+		switch cmd.(type) {
+		// only time the "eval" command
```
Comment (mdlayher) [inline promql/test.go]: Missed one.

File: promql/functions.go [hunk 12]
```diff
@@ -184,6 +184,100 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
...
+// s = computed smoothed values, b = computed trend factors, d = raw
+func calcTrendValue(i int, sf, tf float64, s, b, d []float64) float64 {
+	if i == 0 {
+		return b[0]
+	}
+
+	x := tf * (s[i] - s[i-1])
+	y := (1 - tf) * b[i-1]
+
+	// cache value
```
Comment (mdlayher) [inline promql/functions.go]: Missed one.

**Comment by brian-brazil** [comment]
That's looking good. What are the benchmarks looking like?

**Comment by eliothedeman** [comment]
go test -test.bench=".*" -test.run=none 2>/dev/null
PASS
BenchmarkDoubleSmooth4Week5Min-4	    3000	    507491 ns/op	  894320 B/op	     170 allocs/op
BenchmarkDoubleSmooth1Week5Min-4	   10000	    118521 ns/op	  136593 B/op	      96 allocs/op
BenchmarkDoubleSmooth1Day1Min-4 	   10000	    108947 ns/op	  116586 B/op	     151 allocs/op
ok  	github.com/prometheus/prometheus/promql	4.106s

File: promql/functions_test.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import "testing"
+
+func BenchmarkDoubleSmooth4Week5Min(b *testing.B) {
```
Comment (mdlayher) [inline promql/functions_test.go]: Should be `HoltWinters` instead of `DoubleSmooth`.

File: promql/functions_test.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+eval instant at 4w holt_winters(http_requests[4w], 0.3, 0.3)
+    {path="/foo"} 20160
+`
+
+	bench := NewBenchmark(b, input)
+	bench.Run()
+
+}
+
+func BenchmarkDoubleSmooth1Week5Min(b *testing.B) {
```
Comment (mdlayher) [inline promql/functions_test.go]: Should be `HoltWinters` instead of `DoubleSmooth`.

File: promql/functions_test.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+eval instant at 1w holt_winters(http_requests[1w], 0.3, 0.3)
+    {path="/foo"} 20160
+`
+
+	bench := NewBenchmark(b, input)
+	bench.Run()
+}
+
+func BenchmarkDoubleSmooth1Day1Min(b *testing.B) {
```
Comment (mdlayher) [inline promql/functions_test.go]: Should be `HoltWinters` instead of `DoubleSmooth`.

**Comment by eliothedeman** [comment]
A comparison with the function "changes" running over the same time period as one test.
```
 go test -test.bench=".*" -test.run=none 2>/dev/null
PASS
BenchmarkHoltWinters4Week5Min-4	    3000	    440773 ns/op	  894318 B/op	     170 allocs/op
BenchmarkHoltWinters1Week5Min-4	   10000	    103923 ns/op	  136593 B/op	      96 allocs/op
BenchmarkHoltWinters1Day1Min-4 	   20000	     92933 ns/op	  116586 B/op	     151 allocs/op
BenchmarkChanges1Day1Min-4     	   20000	     69105 ns/op	   79401 B/op	     132 allocs/op
ok  	github.com/prometheus/prometheus/promql	7.641s
```

**Comment by mdlayher** [comment]
No further comments here.  Any last thoughts @brian-brazil ?  Pretty excited to see this go in!

**Comment by brian-brazil** [comment]
Looks like we're more than an order of magnitude faster than when we started out, though there were also a few storage changes in the intervening period which is probably most of it. Still likely a 3X improvement by removing all those if statements. I think this is plenty fast, 50% slower than a trivial function is nothing to complain about for a math-heavy function.

**Comment by brian-brazil** [comment]
That's merged now, thanks!

Will you be doing up the prediction form of this too?

**Comment by eliothedeman** [comment]
Thanks! Yeah I will be doing the prediction version in the next week or so.

**Comment by fabxc** [comment]
Those numbers really look good. Thanks and nice work everyone!

File: promql/bench.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+		t: t,
+	}
+}
+
+// Run runs the benchmark.
+func (b *Benchmark) Run() {
+	b.b.ReportAllocs()
+	b.b.ResetTimer()
+	for i := 0; i < b.b.N; i++ {
+		b.t.RunAsBenchmark(b)
```
Comment (juliusv) [inline promql/bench.go]: While investigating https://github.com/prometheus/prometheus/issues/1743, I noticed that the benchmark error return values are simply ignored here and that the benchmarks fail if you do check the error :-/ That probably means that the benchmarks reported on this PR were invalid?

@eliothedeman would you care fixing these?

**Pull request merged**